File: platforms/dot-net/PulseContribute.DotNet/VFS/Directory.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace PulseContribute.DotNet.VFS
{
    public static class Directory
    {

        public static void ExistsDirectory(string promiseId, string path)
        {
            try
            {
                Promise.Create(promiseId).Resolve(System.IO.Directory.Exists(path));
            }
            catch (Exception)
            {
                Promise.Create(promiseId).resolveFalse();
            }
        }

        public static void CreateDirectory(string promiseId, string path)
        {
            Promise.Create(promiseId).Resolve(
                Helper.TryActionOrFalse(() =>
                {
                    if (!System.IO.Directory.Exists(path))
                    {
                        System.IO.Directory.CreateDirectory(path);
                    }
                }
            ));
        }

        public static void DeleteDirectory(string promiseId, string path)
        {
            try
            {
                Promise.Create(promiseId).Resolve(Helper.TryActionOrFalse(() => {
                    if (System.IO.Directory.Exists(path))
                    {
                        System.IO.Directory.Delete(path);
                    }
                }));
            }
            catch (Exception)
            {
                Promise.Create(promiseId).resolveFalse();
            }
        }

        public static void ReadDirectory(string promiseId, string path)
        {
            List<string> files = new List<string>();
            List<string> dirs = new List<string>();
            
            try
            {
                if (System.IO.Directory.Exists(path))
                {
                    foreach (string file in System.IO.Directory.GetFiles(path))
                    {
                        files.Add(System.IO.Path.GetFileName(file));
                    }
                    foreach (string dir in System.IO.Directory.GetDirectories(path))
                    {
                        dirs.Add(new DirectoryInfo(dir).Name);
                    }
                }            
            } catch (Exception){}
            
            Promise.Create(promiseId).Resolve(new Dictionary<string, object>
            {
                { "dirs", dirs.ToArray() },
                { "files", files.ToArray() }
            });
        }

    }
}

File: api/api/vfs/directory.js
const map = {
  "EXISTS_DIRECTORY": {
    args: [ "STRING" ]
  },
  "CREATE_DIRECTORY": {
    args: [ "STRING" ]
  },
  "DELETE_DIRECTORY": {
    args: [ "STRING" ]
  },
  "READ_DIRECTORY": {
    args: [ "STRING" ],
    after: (val) => {
      if(val){
        if(val.dirs && val.files){
          val.all = val.dirs.concat(val.files);
        }
        return val;
      }
      return { dirs: [], files: [] };
    }
  }
}

const api = {

}

import { presenter } from "../presenter.js";
class Directory extends presenter {

  constructor(path = ""){
    super();
    this.path = path;
  }

  async create(){
    return await this.api.createDirectory(this.path);
  }

  async delete(){
    return await this.api.deleteDirectory(this.path);
  }

  async exists(){
    return await this.api.existsDirectory(this.path);
  }

  async read(){
    return await this.api.readDirectory(this.path);
  }

}

export default { map, api, presenter: Directory }
